bootstrap/kubeadm/api/v1alpha3/kubeadmbootstrapconfig_types.go

```diff
@@ -175,6 +175,98 @@ type File struct {
 	Content string `json:"content"`
 }
 
+// FileSource is a union of all possible external source types for file data.
+// Only one field may be populated in any given instance. Developers adding new
+// sources of data for target systems should add them here.
+type FileSource struct {
+	// Raw represents raw user-provided string content to populate this file.
+	Raw *File `json:"raw,omitempty"`
+
+	// Secret represents a secret that should populate this file.
+	// +optional
+	Secret *SecretFileSource `json:"secret,omitempty"`
+
+	// ConfigMap represents a configMap that should populate this volume
+	// +optional
+	ConfigMap *ConfigMapFileSource `json:"configMap,omitempty"`
+}
+
+type FileSourceOptions struct {
+	// DefaultPermissions specifies optional mode bits to use on created files by default.
+	// Must be a value between 0 and 0777.
+	// +optional
+	DefaultPermissions string `json:"defaultPermissions,omitempty"`
+
+	// DefaultOwner specifies the ownership of the file, e.g. "root:root".
+	// +optional
+	DefaultOwner string `json:"defaultOwner,omitempty"`
+
+	// DefaultEncoding specifies the encoding of the file contents.
+	// +optional
+	DefaultEncoding Encoding `json:"defaultEncoding,omitempty"`
+}
+
+// Adapts a ConfigMap into a FileSource.
+//
+// The contents of the target ConfigMap's Data field will be presented
+// as files using the keys in the Data field as the file names.
+type ConfigMapFileSource struct {
```

Comment:
ConfigMapFileSource and SecretFileSource  are identical, wonder if we can simplify that with a typed object reference?